blog/content/post/2024-03-15-uefi-sb-implementation-mpl.md

```diff
@@ -0,0 +1,361 @@
+    # wic plugin
+    install -d ${DEPLOY_DIR_IMAGE}/bootfiles
+    cp -r ${D}/boot/* ${DEPLOY_DIR_IMAGE}/bootfiles/
+}
+```
+
+We also needed to define the
+[IMAGE_BOOT_FILES](https://docs.yoctoproject.org/singleindex.html#term-IMAGE_BOOT_FILES)
+, which lists files that should
+be installed into the boot partition by the wic tool. We defined it in
+`kas.conf`. Note that the value of this variable heavily depends on your
+system's boot partition layout:
+
+```conf
+IMAGE_BOOT_FILES = " \
+  bootfiles/EFI/BOOT/grubx64.efi;EFI/BOOT/bootx64.efi \
+  bootfiles/EFI/BOOT/grubenv*;EFI/BOOT/
+  bootfiles/EFI/BOOT/grub.cfg*;EFI/BOOT/
+  bootfiles/EFI/BOOT/x86_64-efi/*;EFI/BOOT/x86_64-efi/ \
+  bzImage-initramfs-${MACHINE}.bin;bzImage-initramfs \
+  bzImage-initramfs-${MACHINE}.bin.sig;bzImage-initramfs.sig \
+"
+```
+
+At this point, we encountered several problems with our build.
+
+With `GRUB_SIGN_VERIFY` variable enabled, every grub component needed to be
+signed so that grub could use it. The recipes from `meta-efi-secure-boot`
+take care of generating the signatures. However, the `grubenv` signature was
+missing from our output files. As it turns out, the layer does not automatically
+sign that file. We had to append that feature to the `grub-efi` recipe:
+
+```bitbake
+# grub-efi_%.bbappend
+
+fakeroot python do_sign:append:class-target() {
+    uks_bl_sign(dir + 'grubenv', d)
+}
+
+fakeroot do_chownboot:append() {
+    chown root:root -R "${D}${EFI_BOOT_PATH}/grubenv${SB_FILE_EXT}"
+}
+
+do_deploy:append:class-target () {
+    # deploy missing grubenv.sig file
+    install -m 0600 "${D}${EFI_BOOT_PATH}/grubenv${SB_FILE_EXT}" "${DEPLOYDIR}"
+}
+```
+
+Another issue was that Poky has its own `grub-efi` recipe, where it installs its
+own `grub-efi-grubx64.efi` file. It conflicted with the boot files from
+`meta-efi-secure-boot`, so we had to remove it:
+
+```bitbake
+# grub-efi_%.bbappend
+
+do_deploy:append:class-target () {
+    # remove default grub-efi-grubx64.efi file deployed by .bb from poky; when
+    # it is left in deploydir, bootimg-efi plugin picks it up after cloning
+    # files from IMAGE_BOOT_FILES list
+    # see: https://git.yoctoproject.org/poky/tree/scripts/lib/wic/plugins/source/bootimg-efi.py?id=00c04394cbc5ecaced7cc1bc8bc8787e621f987d#n360
+    rm -rf ${DEPLOYDIR}/${GRUB_IMAGE_PREFIX}${GRUB_IMAGE}
+}
+```
+
+Now your files should be signed automatically during the build. They will be
+deployed along with their signatures. By default, they will be signed using the
+sample keys from
+[meta-signing-key](https://github.com/Wind-River/meta-secure-core/tree/master/meta-signing-key/files/uefi_sb_keys)
+, which is of course extremely unsafe and should only be used for testing.
+`meta-signing-key` provides a
+[script](https://github.com/Wind-River/meta-secure-core/blob/master/meta-signing-key/scripts/create-user-key-store.sh)
+, which generates custom user keys. Such keys should be enrolled in UEFI so
```

Comment:
@PLangowski and for that LockDown.efi was used.. I feel that we already mentioned that in previous section

Reply:
ad87522d2cb6e682dc5e76ebb3c26599f8ec764c